src/Storages/Kafka/StorageKafka.cpp

```diff
@@ -175,22 +175,69 @@ namespace
 
-    /// Configuration prefix
-    const String CONFIG_PREFIX = "kafka";
+    const String CONFIG_KAFKA_TAG = "kafka";
+    const String CONFIG_KAFKA_TOPIC_TAG = "kafka_topic";
+    const String CONFIG_NAME_TAG = "name";
 
-    void loadFromConfig(cppkafka::Configuration & conf, const Poco::Util::AbstractConfiguration & config, const std::string & path)
+    /// Read server configuration into cppkafka configuration, used by global configuration and by legacy per-topic configuration
+    void loadFromConfig(cppkafka::Configuration & kafka_config, const Poco::Util::AbstractConfiguration & config, const String & config_prefix)
     {
-        Poco::Util::AbstractConfiguration::Keys keys;
-        config.keys(path, keys);
+        /// Read all tags one level below <kafka>
+        Poco::Util::AbstractConfiguration::Keys tags;
+        config.keys(config_prefix, tags);
 
-        for (const auto & key : keys)
+        for (const auto & tag : tags)
         {
-            const String key_path = path + "." + key;
-            // log_level has valid underscore, rest librdkafka setting use dot.separated.format
-            // which is not acceptable for XML.
-            // See also https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
-            const String key_name = (key == "log_level") ? key : boost::replace_all_copy(key, "_", ".");
-            conf.set(key_name, config.getString(key_path));
+            if (tag.starts_with(CONFIG_KAFKA_TOPIC_TAG)) ///  Multiple occurrences given as "kafka_topic", "kafka_topic[1]", etc.
+                continue; // used by new per-topic configuration, ignore
+
+            const String setting_path = config_prefix + "." + tag;
+            const String setting_value = config.getString(setting_path);
+
+            // "log_level" has valid underscore, the remaining librdkafka setting use dot.separated.format which isn't acceptable for XML.
+            // See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md
+            const String setting_name_in_kafka_config = (tag == "log_level") ? tag : boost::replace_all_copy(tag, "_", ".");
+            kafka_config.set(setting_name_in_kafka_config, setting_value);
+        }
+    }
+
+    /// Read server configuration into cppkafa configuration, used by new per-topic configuration
+    void loadTopicConfig(cppkafka::Configuration & kafka_config, const Poco::Util::AbstractConfiguration & config, const String & config_prefix, const String & topic)
+    {
+        /// Read all tags one level below <kafka>
+        Poco::Util::AbstractConfiguration::Keys tags;
+        config.keys(config_prefix, tags);
+
+        for (const auto & tag : tags)
+        {
+            /// Consider tag <kafka_topic>. Multiple occurrences given as "kafka_topic", "kafka_topic[1]", etc.
+            if (tag.starts_with(CONFIG_KAFKA_TOPIC_TAG))
```

Comment:
early continue if it doesn't start with `CONFIG_KAFKA_TOPIC_TAG` to eliminate indentation

Reply:
Done